Add information about clip, translate, rendering and refresh strategy. (#933)
MicroEJ/docs: ApplicationDeveloperGuide/UI/MicroUI/rendering.rst

.. _GraphicsContext.setTranslation(int, int): https://repository.microej.com/javadoc/microej_5.x/apis/ej/microui/display/GraphicsContext.html#setTranslation-int-int-
.. _GraphicsContext.translate(int, int): https://repository.microej.com/javadoc/microej_5.x/apis/ej/microui/display/GraphicsContext.html#translate-int-int-
.. _GraphicsContext.resetTranslation(): https://repository.microej.com/javadoc/microej_5.x/apis/ej/microui/display/GraphicsContext.html#resetTranslation--
.. _GraphicsContext.getTranslationX(): https://repository.microej.com/javadoc/microej_5.x/apis/ej/microui/display/GraphicsContext.html#getTranslationX--
.. _GraphicsContext.getTranslationY(): https://repository.microej.com/javadoc/microej_5.x/apis/ej/microui/display/GraphicsContext.html#getTranslationY--

Rendering Pipeline
------------------

The Rendering Pipeline of a MicroUI application consists of two main phases: Render, and Flush.

1. *Render*: During this phase, the application executes its rendering code to perform the necessary drawing operations.
   The application can either draw the whole screen for each frame or just a part.

> 2. *Flush*: This phase involves copying the UI working buffer to the screen buffer.

Reviewer: ```suggestion
2. *Flush*: This phase involves copying the back buffer to the front buffer.
```

   The VEE Port performs this operation, and it is the responsibility of the VEE Port developer to optimize this process, for example, by utilizing a GPU.

To reduce the number of pixels drawn, it may be interesting to redraw only subpart(s) of the display.
In this case, the *Render* phase consists in setting the clip to the desired region and then draw inside this region.
And again for each region to redraw.

Display Buffer Management
-------------------------

The display usually involves several buffers.
It may be for one or several of these reasons:

- the display has its own memory that is not mapped by the CPU,
- to avoid seeing the steps of the drawing (if the display reads in the same memory as the application is drawing to),
- to allow the application to draw more quickly after a flush,
- etc.

More information can be found in the :ref:`section_display` section of the VEE Porting Guide.

When the buffer used to draw is changed after a flush, the new buffer needs to be "refreshed" with the content of the previous buffer to avoid glitches.
This copy is managed by the Buffer Refresh Strategy (BRS).
To save some copy time, the BRS optimizes the refreshed region.

> For example if the full screen is drawn nothing needs to be restored (the same goes if the very same zone is drawn several times in a row).

Reviewer: ```suggestion
For example, if the full screen is drawn, nothing needs to be restored (the same goes if the very same zone is drawn several times in a row).
```

More information can be found in the :ref:`section_brs` section of the VEE Porting Guide.

The BRS has two ways to detect the drawn regions.
Both are using the clip:

1. The first one does not require any action from the application, the BRS saves the clip when a drawing is done.
2. The second one consists in declaring the list of the regions that will be modified before starting the drawings.
   This is done by setting the clip and calling the `GraphicsContext.notifyDrawingRegion()`_ method for each region.
   Declaring the regions is advised when redrawing several parts of the display in the frames.

.. _GraphicsContext.notifyDrawingRegion(): https://repository.microej.com/javadoc/microej_5.x/apis/ej/microui/display/GraphicsContext.html#notifyDrawingRegion--

> Since the refreshing is based on the clip, it means that all the pixels in the clips need to be drawn since they will not be refreshed.

Reviewer: ```suggestion
Since the refreshing is based on the clip, it means that all the pixels in the clip need to be drawn since they will not be refreshed.
```

> In other words, the unchanged pixels inside the clip may contain unexpected data that will create glitches in the display.

Reviewer: ```suggestion
In other words, the pixels left unchanged inside the clip may contain unexpected data that will create glitches in the display.
```
